components/script/dom/xoriginwindow.rs

```diff
@@ -0,0 +1,141 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::codegen::Bindings::XOriginWindowBinding;
+use dom::bindings::codegen::Bindings::XOriginWindowBinding::XOriginWindowMethods;
+use dom::bindings::js::{JS, MutNullableJS, Root};
+use dom::bindings::reflector::DomObject;
+use dom::bindings::str::DOMString;
+use dom::browsingcontext::BrowsingContext;
+use dom::globalscope::GlobalScope;
+use dom::xoriginlocation::XOriginLocation;
+use ipc_channel::ipc;
+use js::jsapi::{JSContext, HandleValue};
+use js::jsval::{JSVal, UndefinedValue};
+use msg::constellation_msg::PipelineId;
+
+/// Represents a dissimilar-origin `Window` that exists in another script thread.
+///
+/// Since the `Window` is in a different script thread, we cannot access it
+/// directly, but some of its accessors (for example `window.parent`)
+/// still need to function.
+///
+/// In `browsingcontext.rs`, we create a custom window proxy for these windows,
+/// that throws security exceptions for most accessors. This is not a replacement
+/// for XOWs, but provides belt-and-braces security.
+
+#[dom_struct]
+pub struct XOriginWindow {
+    /// The global for this window.
+    globalscope: GlobalScope,
+
+    /// The browsing context this window is part of.
+    browsing_context: JS<BrowsingContext>,
+
+    /// The location of this window, initialized lazily.
+    location: MutNullableJS<XOriginLocation>,
+}
+
+impl XOriginWindow {
+    #[allow(unsafe_code)]
+    pub fn new(browsing_context: &BrowsingContext) -> Root<XOriginWindow> {
+        let globalscope = browsing_context.global();
+        let cx = globalscope.get_cx();
+        // Any timer events fired on this window are ignored.
+        let (timer_event_chan, _) = ipc::channel().unwrap();
+        let win = box XOriginWindow {
+            globalscope: GlobalScope::new_inherited(PipelineId::new(),
```

Comment:
We may want a TODO here about storing the actual pipeline that this cross-origin window represents.